Added implementation for KMeans algorithm in GIL

Sayan-Chaudhuri opened the pull request:
This is an implementation of Kmeans Clustering algorithm for Boost GIL. The input image filename(absolute/relative) must be passed as a command line argument.During execute ,input the required number of iterations and K value that indicates number of clusters.
Output is stored in the file named output-kmeans.tif
I have attached the example output of my implementation for the input image frog.jpg(attached),and number of clusters and iterations being 10 and 1 respectively.
I would request the community to kindly provide feedback for my implementation
I would also like to use the implementation of this algorithm for my gil competency test for GSOC 2021.

<!-- Pull Requests MUST come from topic branch based on develop, and NEVER on `master) --->

### Description

<!-- What does this pull request do? -->

### References

<!-- Any links related to this PR: issues, other PRs, mailing list threads, StackOverflow questions, etc. -->

### Tasklist

<!-- Add YOUR OWN TASK(s), especially if your PR is a work in progress -->

- [ ] Add test case(s)
- [ ] Ensure all CI builds pass
- [ ] Review and approve

Sayan-Chaudhuri committed:
Added implementation for KMeans algorithm in GIL

This is an implementation of Kmeans Clustering algorithm for Boost GIL. The input image filename(absolute/relative) must be passed as a command line argument.During execute ,input the required number of iterations and K value that indicates number of clusters.
Output is stored in the file named output-kmeans.tif
I have attached the example output of my implementation for the input image frog.jpg(attached),and number of clusters and iterations being 10 and 1 respectively.
I would request the community to kindly provide feedback for my implementation
I would also like to use the implementation of this algorithm for my gil competency test for GSOC 2021.

lpranam commented:
This algorithm is used everywhere so shouldn't this be introduced as a usable functionality of the library instead of an example?

lpranam reviewed (changes requested):
we don't have to provide output file

Sayan-Chaudhuri commented:
I had thought that first my implementation will be checked whether its working correctly or not and so I did it in the form of an example. But if you suggest, I shall make it in the form of a usable functionality and then raise a pull request. I am also removing the output files following your suggestion.

lpranam commented:
First example and then the actual implementation just increases the amount of work reviewers and you will have to do. Let's just review the final thing.

Sayan-Chaudhuri committed:
Delete frog.jpg

Sayan-Chaudhuri committed:
Deleted output images

Sayan-Chaudhuri committed:
.cpp file to be changed into .h for review

Sayan-Chaudhuri committed:
Added KMeans Clustering functionality for Boost 

Turned the example file into a usable functionality(.hpp) for KMeans Clustering as suggested by lPranam and added it to the image_processing folder of boost.

Sayan-Chaudhuri commented:
@lpranam  I have made the necessary changes as suggested .

lpranam commented:
new functionality also needs tests...

Sayan-Chaudhuri commented:
@lpranam Ok,I shall soon make the changes and update here

Sayan-Chaudhuri commented:
@lpranam  Is it ok if for the test file I keep a static dataset? I saw the test files of other algorithms where they have fixed the input and expected output data like the pixel value of images

lpranam commented:
@Sayan-Chaudhuri yes, as far as it covers the cases it is okay.

Sayan-Chaudhuri commented:
@lpranam I wish to upload the test file along with the KMeans implementation but I want to clarify certain things.
1. I have used a  single dataset generated with the help of  make_blobs() function in python. I have tested it against the existing implementations of kmeans with random centre intialization like in sklearn and OPENCV. So, for benchmarking I have used the silhoutte score obtained using those implementations on that dataset.For a set of 20 runs of the algorithm using different centre initializations, I obtain a silhoutte score of  above 0.8 for 90% of the runs .Using the existing implementations in sklearn and OPENCV , this score has come to be 0.83 with the same number of runs. Is it OK If I submit my implementation then?
2. I have implemented the entire silhoutte score algorithm while testing my implementation. Silhoutte score is very common in clustering techniques and Boost does not have a separate implementation for that. So should I make a separate header file for the Silhoutte score implementation so that it can be reviewed and later merged with the library if deemed ok? I have already googled and I found no api for Silhoutte score calculation in boost.

If you can kindly find time to clarify these doubts, I will be highly grateful.

Sayan-Chaudhuri commented:
Upon your clarification, I shall push the files accordingly

lpranam commented:
you should upload it if, it does not fit than obviously can be removed but need to have look.

lpranam commented:
can also you PR must pass all the CI checks

Sayan-Chaudhuri committed:
Delete KMeans.hpp

Sayan-Chaudhuri committed:
Change in function parameters,namespace

Sayan-Chaudhuri commented:
Added a new header file for Kmeans

Sayan-Chaudhuri commented:
@lpranam

Sayan-Chaudhuri committed:
Added test file as suggested

Sayan-Chaudhuri commented:
@lpranam I have also added the test file

meshtag commented:
Hi @Sayan-Chaudhuri ,
I had a look at [this](https://github.com/boostorg/gil/pull/587/checks?check_run_id=2244613275) build which contains the following message

> error: toolset gcc initialization:
> error: version '8' requested but 'g++-8' not found and version '7.5.0' of default 'g++' does not match> 

I don't think this is related to any changes made by you in this PR, you should probably update this branch with latest develop branch of boost Gil.
Pushing again after updating should probably solve this issue.

PS : I encountered a similar error [here](https://github.com/boostorg/gil/pull/562#issuecomment-811979511)

Sayan-Chaudhuri commented:
@meshtag how to do so what you have mentioned?

meshtag commented:
There are a couple of ways to do this, you can look [here](https://stackoverflow.com/questions/3903817/pull-new-updates-from-original-github-repository-into-forked-github-repository) to understand some common ones.

mloskot commented:
@Sayan-Chaudhuri 
>how to do so what you have mentioned?

Please, read the CONTRIBUTING.md on updating your PR.

Updating PR on GitHub is a common operation, so it is very well documented on the web, GitHub docs, StackOverflow.
